README

         Text::Delimited - Delimited Text Processing Module - VERSIONTAG

================================================================================
SYNOPSIS:
================================================================================

    Text::Delimited provides a programattical interface to data stored in
    delimited text files. It is dependant upon the first row of the text
    file containing header information for each corresponding "column" in
    the remainder of the file.

    After instancing, for each call to Read the next row's data is returned
    as a hash reference. The individual elements are keyed by their
    corresonding column headings.

================================================================================
INSTALLATION:
================================================================================

    To install this module type the following:

       perl Makefile.PL
       make
       make test
       make install

================================================================================
USAGE:
================================================================================

    A short example of usage is detailed below. It opens a pipe delimited
    file called 'infile.txt', reads through every row and prints out the
    data from "COLUMN1" in that row. It then closes the file.

      my $file = new Text::Delimited;
      $file->delimiter('|');
      $file->open('infile.txt');

      my @header = $file->fields;

      while ( my $row = $file->read ) {
        print $row->{COLUMN1}, "\n";
      }

      $file->close;

    The close() method is atuomatically called when the object passes out of
    scope. However, you should not depend on this. Use close() when
    approrpiate.

    Other informational methods are also available. They are listed blow:

================================================================================
METHODS:
================================================================================

    close()

    Closes the file or connection, and cleans up various bits.

    delimiter(delimiter)

    Allows you to set the delimiter if a value is given. The default
    delimiter is a tab. Returns the delimiter.

    fields()

    Returns an array (or arrayref, depending on the requested context) with
    the column header fields in the order specified by the source file.

    filename()

    If Open was given a filename, this function will return that value.

    linenumber()

    This returns the line number of the last line read. If no calls to Read
    have been made, will be 0. After the first call to Read, this will
    return 1, etc.

    new([filename|filepointer],[enumerate])

    Creates a new Text::Delimited object. Takes optional parameter that is
    either a filename or a globbed filehandle. Files specified by filename
    must already exist.

    Can optionally take a second argument. If this argument evaluates to
    true, Text::Delimited will append a _NUM to the end of all fields with
    duplicate names. That is, if your header row contains 2 columns named
    "NAME", one will be changed to NAME_1, the other to NAME_2.

    open([filename|filepointer], [enumerate])

    Opens the given filename or globbed filehandle and reads the header
    line. Returns 0 if the operation failed. Returns the file object if
    succeeds.

    Can optionally take a second argument. If this argument evaluates to
    true, Text::Delimited will append a _NUM to the end of all fields with
    duplicate names. That is, if your header row contains 2 columns named
    "NAME", one will be changed to NAME_1, the other to NAME_2.

    read()

    Returns a hashref with the next record of data. The hash keys are
    determined by the header line.

    __DATA__ and __LINE__ are also returned as keys.

    __DATA__ is an arrayref with the record values in order.

    __LINE__ is a string with the original tab-separated record.

    This method returns undef if there is no more data to be read.

    setmode(encoding)

    Set the given encoding scheme on the input file to allow for reading files
    encoded in standards other than ASCII.

================================================================================
EXPORTABLE METHODS:
================================================================================

    For convienience, the following methods are exportable. These are handy
    for quickly writing output delimited files.

    d_join(@STUFF)
    Delimited Join. Returns the given array as a string joined with the 
    current delimiter.

    d_line(@STUFF)
    Delimited Line. Returns the given array as a string joined with the 
    current delimiter and with newline appended.

================================================================================
AUTHORSHIP:
================================================================================

    Text::Delimited VERSIONTAG DATETAG

    (c) 2004-YEARTAG, Phillip Pollard <bennie@cpan.org>
    Released under the Perl Artistic License

    I'd like to thank PetBlvd for sponsoring continued work on this module.
    http://www.petblvd.com/

    Additional contributions by Kristina Davis <krd@menagerie.tf>
    Based upon the original module by Andrew Barnett <abarnett@hmsonline.com>

    Derived from Util::TabFile 1.9 2003/11/05
    With permission granted from Health Market Science, Inc.


Changes
Revision history for Perl extension Text::Delimited.

VERSIONTAG  DATETAG

	Using GNU TAR now. The previous release had PAX headers in the tar file.
	Build scripts now automatically configure minium perl version.

2.10  2014/03/26

	Including credit for the RT on the 2.00 change info

2.09  2014/03/07

	Now using warnings.

2.08  2014/03/03

	Automated build, now properly constructing meta data.

	Corrected POD errors

	Adding a POD test

2.07  2014/02/24

	Meta information updated
	
	Added LICENSE file to distribution

	Tests updated to Test::Simple.

	Version leap to match automated build numbers.
        
2.00  2006/07/31

	Provided non-capitalized method names to match Perl style as suggested
         by CPAN rating comments. Historic "Java-style" method names still
	 function.

	Corrected POD and docs for new method names, version changes, other 
	 grammatical errors and removed accidental references to TabFile.pm

	Added initialization checks on exported methods to avoid superfluous
	 warnings generated when -w is used.

	Handle blank lines at the end of a file gracefully as per RT #12006

	Added credit to PetBlvd (http://www.petblvd.com) for the sponsorship
	 of further development.

1.93  2004/07/21

	Removed autovivification of the internal filehandle. Now using the
	 Symbol::gensym() to be backwards compatible to earlier versions
	 of Perl 5. Thanks to Shrinivas Potnuru for the bug report.

1.92  2004/04/02

	Full copyright and license information.

1.91  2004/01/15

	Derived from HMS tool Util::TabFile (1.9 2003/11/05)

	Support for changing delimiter added.

	Module made to pass strict.


build.pl
#!/usr/bin/env perl

use Data::Dumper;
use strict;

# These will be necessary for Make Maker to make the module with proper files.
require CPAN::Meta;
require CPAN::Meta::Converter;
require CPAN::Meta::YAML;
require ExtUtils::MakeMaker;

print "Using ExtUtils::MakeMaker ($ExtUtils::MakeMaker::VERSION)\n";
print "Using CPAN::Meta ($CPAN::Meta::VERSION)\n";

die "You need a modern version of ExtUtils::MakeMaker." unless $ExtUtils::MakeMaker::VERSION > 6;

### Config

my ($module, $author,  $license, $abstract, $description, $perl_ver, %requires);

open INFILE, '<', 'config.txt' or die "Can't open 'config.txt'";
my $text = join('',<INFILE>);
close INFILE;

eval $text;
die $@ if $@;

die "Bad config." unless $module && $author && $license && 
  $abstract && $description && $perl_ver && %requires;

### Post config

my $path_chunk = $module;
$path_chunk =~ s/::/-/g;

my $bug  = 'https://rt.cpan.org/Dist/Display.html?Name='.$path_chunk;
my $repo = 'http://github.com/bennie/perl-' . $path_chunk;
my $git  = 'git://github.com/bennie/perl-'.$path_chunk.'.git';

my $sourcefile = 'lib/' . $module . '.pm';
$sourcefile =~ s/::/\//g;

my $require_text = Dumper(\%requires);
$require_text =~ s/\$VAR1 = //;
$require_text =~ s/;$//;
1 while chomp($require_text);

### Bump the version if asked

for my $arg (@ARGV) {
  version_bump() and last if $arg eq '--bump';
  if ( $arg eq '--retag'   ) { version_retag();  exit; }
  if ( $arg eq '--version' ) { print version_current(), "\n"; exit; }
}

### External data

my $version = version_current();
my $date    = `date '+%Y/%m/%d'`;
my $year    = `date '+%Y'`;
my $distdir = $path_chunk .'-' . $version;

chomp $date;
chomp $year;
chomp $distdir;

print "
Version : $version
Date    : $date
Year    : $year
Dist    : $distdir

";

### Write Makefile.PL

open MAKEFILE, '>', 'Makefile.PL';

print MAKEFILE "use ExtUtils::MakeMaker;

WriteMakefile(
  ABSTRACT => \"$abstract\",
  AUTHOR   => '$author',
  LICENSE  => '$license',
  NAME     => '$module',
  VERSION  => '$version',

  PREREQ_PM => $require_text,

  ( \$ExtUtils::MakeMaker::VERSION < 6.46
        ? ()
        : ( META_MERGE => {
                'meta-spec' => { version => 2 },
                no_index => {directory => [qw/t/]},
                provides => {
                     '$module' => {
                          file    => '$sourcefile',
                          version => '$version'
                     }
                },
                requires  => {perl => '$perl_ver'},
                resources => {
                    repository => {
                        type => 'git',
                        url  => '$git',
                        web  => '$repo',
                    },
                    bugtracker => {
                        web => '$bug',
                    },

                },
            },
        )
    )

);";

close MAKEFILE;

### Build the distribution directory

print `perl Makefile.PL`;
print `make distmeta`;

### Build META.json

unless ( -f $distdir.'/META.json' ) {
  my $distmeta = {

    # Required
    abstract => $abstract,
    author   => [ $author ],
    license  => [ $license ],
    name     => $path_chunk,
    version  => $version,

    # optional
    dynamic_config => 1,
    'meta-spec' => { version => '2', url => 'http://search.cpan.org/perldoc?CPAN::Meta::Spec' },
    generated_by => "CPAN::Meta version $CPAN::Meta::VERSION",

    # 2.0 only stuff
    description =>  $description,
    release_status => 'stable',

    prereqs => {
      runtime => {
        requires => \%requires,
        recommends => { },
      },
      build => {
        requires => \%requires,
      }
    },
    resources => {
      license    => [ 'http://dev.perl.org/licenses/' ],
      bugtracker => { web => $bug },
      repository => { web => $repo , type => 'git', url => $git },
    },
  };

  $distmeta->{prereqs}->{runtime}->{requires}->{perl} = $perl_ver;

  my $meta = CPAN::Meta->create($distmeta);
  print "Generating META.json on my own.\n";
  $meta->save($distdir.'/META.json');

  print "Adding META.json to the MANIFEST\n";
  open MANIFEST, '>>', $distdir.'/MANIFEST';
  print MANIFEST "META.json\nMakefile.PL";
  close MANIFEST;

}

### Updating the tags

print  "\nUpdating DATETAG -> $date\n";
system "find $distdir -type f | xargs perl -p -i -e 's|DATETAG|$date|g'";
print  "Updating VERSIONTAG -> $version\n";
system "find $distdir -type f | xargs perl -p -i -e 's|VERSIONTAG|$version|g'";
print  "Updating YEARTAG -> $year\n";
system "find $distdir -type f | xargs perl -p -i -e 's|YEARTAG|$year|g'";
print "\n";

### Build the tarball

unlink($distdir.'.tar')    if -f $distdir.'.tar';
unlink($distdir.'.tar.gz') if -f $distdir.'.tar.gz';

system "tar cvf $distdir.tar $distdir && gzip --best $distdir.tar";

### Cleanup

unlink('Makefile');
unlink('Makefile.old');
unlink('Makefile.PL');
unlink('MYMETA.json');
unlink('MYMETA.yml');

print "\nDONE!\n";

### Subroutines

sub version_bump {
  my $version = version_current();
  my $next_version = version_next($version);
  print "Bumping version from $version to $next_version\n";
  system("git tag -a $next_version -m 'Version $next_version'");
  system("git push --tags origin master");
  system("git push --tags github master");
  our $version = $next_version;
}

sub version_current {
  our $version;
  return $version if defined $version and length $version;
  $version = `git describe --long | tr '-' ' ' | awk '{ print \$1 }'`;
  chomp $version;
  $version = '0.01' unless length $version;
  return $version;
}

sub version_next {
  my $old = shift @_;
  $old =~ /^(.+)\.(.+?)$/;
  my ($main,$sub) = ($1,$2);
  my $length = length($sub);
  $sub++;
  if ( length($sub) > $length ) {
    warn "WARNING: Version length now larger! (From $length digits to ", length($sub), ")\n";
    $length = length($sub);
  }
  return sprintf('%s.%0'.$length.'d',$main,$sub);
}

sub version_retag {
  my $tag = version_current();
  system("git tag -d $tag");
  system("git push origin :refs/tags/$tag");
  system("git push github :refs/tags/$tag");
  print "\nTagging current as $tag\n";
  system("git tag -a $tag -m 'Version $tag'");
  system("git push --tags origin master");
  system("git push --tags github master");
}


build.sh
#!/bin/sh
set -e

VERSION=`./version.pl`
BUILD=`./version.pl --build`
DATE=`date '+%Y/%m/%d'` 
YEAR=`date '+%Y'`
TARDIR="Text-Delimited-$VERSION";

echo
echo "Version : $VERSION"
echo "Build   : $BUILD"
echo "Date    : $DATE"
echo "Year    : $YEAR"
echo "Tar Dir : $TARDIR"
echo

if [ -d build ];
  then echo "Cleaning Build directory:"; rm -rfv build; echo;
fi

echo "Creating the build directory:"
mkdir -pv "build/$TARDIR"
echo

echo "Copying files"
rsync -av --files-from=MANIFEST ./ "build/$TARDIR/"
echo

echo "Updating date tags."
find build -type f | xargs perl -p -i -e "s|DATETAG|$DATE|g"
echo

echo "Updating version tags."
find build -type f | xargs perl -p -i -e "s|VERSIONTAG|$VERSION|g"
echo

echo "Updating build tags."
find build -type f | xargs perl -p -i -e "s|BUILDTAG|$BUILD|g"
echo

echo "Updating date tags."
find build -type f | xargs perl -p -i -e "s|YEARTAG|$YEAR|g"
echo

echo "Building the tar file."
cd build && tar czvf $TARDIR.tar.gz $TARDIR
echo

git push

echo DONE!


lib/Text/Delimited.pm
=head1 NAME

Text::Delimited - Module for parsing delimited text files

=head1 SYNOPSIS

Text::Delimited provides a programattical interface to data stored in 
delimited text files. It is dependant upon the first row of the text 
file containing header information for each corresponding "column" in the 
remainder of the file.

After instancing, for each call to Read the next row's data is returned as 
a hash reference. The individual elements are keyed by their corresonding 
column headings.

=head1 USAGE

A short example of usage is detailed below. It opens a pipe delimited file 
called 'infile.txt', reads through every row and prints out the data from 
"COLUMN1" in that row. It then closes the file.

  my $file = new Text::Delimited;
  $file->delimiter('|');
  $file->open('infile.txt');

  my @header = $file->fields;

  while ( my $row = $file->read ) {
    print $row->{COLUMN1}, "\n";
  }

  $file->close;

The close() method is atuomatically called when the object passes out of 
scope. However, you should not depend on this. Use close() when 
approrpiate.

Other informational methods are also available. They are listed blow:

=head1 METHODS:

=over

=item close()

Closes the file or connection, and cleans up various bits.

=item delimiter(delimiter)

Allows you to set the delimiter if a value is given. The default 
delimiter is a tab. Returns the delimiter.

=item fields()

Returns an array (or arrayref, depending on the requested context) with 
the column header fields in the order specified by the source file.

=item filename()

If open() was given a filename, this function will return that value.

=item linenumber()

This returns the line number of the last line read. If no calls to Read 
have been made, will be 0. After the first call to Read, this will return 
1, etc.

=item new([filename|filepointer],[enumerate])

Creates a new Text::Delimited object. Takes optional parameter that is either
a filename or a globbed filehandle. Files specified by filename must 
already exist.

Can optionally take a second argument. If this argument evaluates to true,
Text::Delimited will append a _NUM to the end of all fields with duplicate names.
That is, if your header row contains 2 columns named "NAME", one will be 
changed to NAME_1, the other to NAME_2.

=item open([filename|filepointer], [enumerate])

Opens the given filename or globbed filehandle and reads the header line. 
Returns 0 if the operation failed. Returns the file object if succeeds.

Can optionally take a second argument. If this argument evaluates to true,
Text::Delimited will append a _NUM to the end of all fields with duplicate names.
That is, if your header row contains 2 columns named "NAME", one will be 
changed to NAME_1, the other to NAME_2.

=item read()

Returns a hashref with the next record of data. The hash keys are determined
by the header line. 

__DATA__ and __LINE__ are also returned as keys.

__DATA__ is an arrayref with the record values in order.

__LINE__ is a string with the original tab-separated record. 

This method returns undef if there is no more data to be read.

=item setmode(encoding)

Set the given encoding scheme on the input file to allow for reading files
encoded in standards other than ASCII.

=back

=head1 EXPORTABLE METHODS

For convienience, the following methods are exportable. These are handy 
for quickly writing output delimited files.

=over

=item d_join(@STUFF)

Delimited Join. Returns the given array as a string joined with the
current delimiter.

=item d_line(@STUFF)

Delimited Line. Returns the given array as a string joined with the
current delimiter and with newline appended.

=back

=head1 BUGS AND SOURCE

	Bug tracking for this module: https://rt.cpan.org/Dist/Display.html?Name=Text-Delimited

	Source hosting: http://www.github.com/bennie/perl-Text-Delimited

=head1 VERSION

    Text::Delimited vVERSIONTAG (DATETAG)

=head1 COPYRIGHT

    (c) 2004-YEARTAG, Phillip Pollard <bennie@cpan.org>

=head1 LICENSE

This source code is released under the "Perl Artistic License 2.0," the text of
which is included in the LICENSE file of this distribution. It may also be
reviewed here: http://opensource.org/licenses/artistic-license-2.0

=head1 AUTHORSHIP

    I'd like to thank PetBlvd for sponsoring continued work on this module.
    http://www.petblvd.com/

    Additional contributions by Kristina Davis <krd@menagerie.tf>
    Based upon the original module by Andrew Barnett <abarnett@hmsonline.com>

    Originally derived from Util::TabFile 1.9 2003/11/05
    With permission granted from Health Market Science, Inc.

=cut

package Text::Delimited;

use Symbol;

use 5.006001;
use warnings;
use strict;

$Text::Delimited::VERSION = 'VERSIONTAG';

### Private mthods

sub DESTROY {   
  return $_[0]->Close;
}

sub _line {
  my $self = shift @_;
  $self->{CURRENT_LINE} = readline($self->{FP});
  $self->{CURRENT_LINE} =~ s/[\r\n]+$//;
  $self->{CURRENT_DATA} = [ split /\Q$self->{DELIMITER}\E/o, $self->{CURRENT_LINE} ];
  $self->{LINE_NUMBER}++;
  return $self->{CURRENT_DATA};
}

### Public Methods

sub Close {
  my $self = shift @_;
  return $self->close(@_);
}

sub close {
  my $self = shift @_;
  close $self->{FP} if $self->{FP};

  $self->{CURRENT_DATA} = $self->{CURRENT_LINE} = $self->{FILENAME} = 
  $self->{FP} = $self->{HDR} = $self->{LINE_NUMBER} = undef;

  return 1;
}

sub Delimiter {
  my $self = shift @_;
  return $self->delimiter(@_);
}

sub delimiter {
  my $self = shift @_;
  my $new  = shift @_;

  if ( $new and $self->{LINE_NUMBER} > 0 ) {
    warn "You cannot change the delimiter after you have opened the file for processing.\n";
  } elsif ( $new ) {
    $self->{DELIMITER} = $new;
  }
    
  return $self->{DELIMITER};
}

sub Fields {
  my $self = shift @_;
  return $self->fields(@_);
}

sub fields {
  my $self = shift @_;
  return wantarray ? @{$self->{HDR}} : $self->{HDR};
}

sub FileName {
  my $self = shift @_;
  return $self->filename(@_);
}

sub filename {
  return $_[0]->{FILENAME};
}

sub LineNumber {
  my $self = shift @_;
  return $self->linenumber(@_);
}

sub linenumber {
  return $_[0]->{LINE_NUMBER};
}

sub new {
  my $class = shift @_;
  my $file  = shift @_;
  my $enumerate = shift @_;

  my $self = {
    DELIMITER   => "\t",
    LINE_NUMBER => 0
  };
    
  bless $self, $class;

  $self->_init;

  my $status = $self->open($file, $enumerate) if $file;
  
  return $self;
}

sub _init { }

sub Open {
  my $self = shift @_;
  return $self->open(@_);
}

sub open {
    my $self = shift @_;
    my $file = shift @_;
    my $enumerate = shift @_;

    $self->{ENUMERATE} = $enumerate;

    if ( ref($file) eq 'GLOB' || not $file ) {
      $self->{FP} = $file || \*STDIN;
      $self->{FILENAME} = 'GLOB';
    } elsif ( -r $file ) {
      $self->{FP} = gensym;
      open $self->{FP}, $file or die "Can't open the file $file\n";
      $self->{FILENAME} = $file;
    } else {
      die "$file is neither a filehandle or an existing, readable file.";
    }

    $self->{HDR} = $self->_line;
  
    my %fields = ( );
    my %dupes = ( );
    for (my $i = 0; $i < scalar @{$self->{HDR}}; $i++) {
	my $field = ${$self->{HDR}}[$i];
	if ($fields{$field}) {
	    if ($self->{ENUMERATE} > 0) {
		$dupes{$field} += 1;
		${$self->{HDR}}[$i] = "$field\_$dupes{$field}";
	    }
	    else {
		die "ERROR: There is a duplicate column name: $field.  This is not good.\n";
	    }
	}
	$fields{$field} = 1;
    }
    return 1;
}

sub Read {
  my $self = shift @_;
  return $self->read(@_);
}

sub read {
  my $self = shift @_;
  my $out  = {};

  my $data = $self->_line;

  return undef unless scalar @$data > 0;

  my $i = 0;
  for my $val ( @$data ) {
    $val =~ s/^\s+|\s+$//g;
    $out->{$self->{HDR}->[$i++]} = $val;
  }

  $out->{__DATA__} = $self->{CURRENT_DATA};
  $out->{__LINE__} = $self->{CURRENT_LINE};

  return $out;
}

sub setMode {
  my $self = shift @_;
  return $self->setMode(@_);
}

sub setmode {
  my $self = shift @_;
  my $mode = shift @_;
  return binmode $self->{FP}, $mode;
}

### Exportable methods

sub d_join {
  if ( ref($_[0]) ) {
    my $self = shift @_;
    return join($self->{DELIMITER},map {defined($_)?$_:''} @_);
  } else {
    return join("\t",map {defined($_)?$_:''} @_);
  }
}

sub d_line {
  if ( ref($_[0]) ) {
    my $self = shift @_;
    return join($self->{DELIMITER},map {defined($_)?$_:''} @_) . "\n";
  } else {
    return join("\t",map {defined($_)?$_:''} @_) . "\n";
  }
}

1;
